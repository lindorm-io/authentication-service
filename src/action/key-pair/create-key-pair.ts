import Joi from "@hapi/joi";
import { IAuthContext } from "../../typing";
import { KeyType } from "@lindorm-io/key-pair";
import { assertAccountAdmin, generateKeyPair } from "../../support";

export interface ICreateKeyPairOptions {
  type: KeyType;
}

export interface ICreateKeyPairData {
  algorithm: string;
  keyPairId: string;
  type: string;
}

const schema = Joi.object({
  type: Joi.string().valid(KeyType.EC, KeyType.RSA).required(),
});

export const createKeyPair = (ctx: IAuthContext) => async (
  options: ICreateKeyPairOptions,
): Promise<ICreateKeyPairData> => {
  await schema.validateAsync(options);

  const { cache, logger, repository } = ctx;
  const { type } = options;

  await assertAccountAdmin(ctx)();

  const keyPair = await generateKeyPair(type);

  await repository.keyPair.create(keyPair);
  await cache.keyPair.create(keyPair);

  logger.debug("key pair created", {
    algorithm: keyPair.algorithm,
    id: keyPair.id,
    type: keyPair.type,
  });

  return {
    algorithm: keyPair.algorithm,
    keyPairId: keyPair.id,
    type: keyPair.type,
  };
};
