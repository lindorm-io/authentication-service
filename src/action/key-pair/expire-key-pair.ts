import Joi from "@hapi/joi";
import { IKoaAuthContext } from "../../typing";
import { InvalidExpiryString } from "../../error";
import { add } from "date-fns";
import { assertAccountAdmin } from "../../support";
import { stringToDurationObject, stringToMilliseconds } from "@lindorm-io/core";

export interface IExpireKeyPairOptions {
  keyPairId: string;
  expires: string;
}

const schema = Joi.object({
  keyPairId: Joi.string().guid().required(),
  expires: Joi.string().required(),
});

export const expireKeyPair = (ctx: IKoaAuthContext) => async (options: IExpireKeyPairOptions): Promise<void> => {
  await schema.validateAsync(options);

  const { cache, logger, repository } = ctx;
  const { expires, keyPairId } = options;

  await assertAccountAdmin(ctx)();

  if (stringToMilliseconds(expires) === 0) {
    throw new InvalidExpiryString(expires);
  }

  const keyPair = await repository.keyPair.find({ id: keyPairId });

  keyPair.expires = add(new Date(), stringToDurationObject(expires));

  const updated = await repository.keyPair.update(keyPair);

  try {
    await cache.keyPair.update(updated);
  } catch (err) {
    logger.error("could not update cache", err);
  }

  logger.debug("key pair updated", {
    id: keyPair.id,
    expires: keyPair.expires,
  });
};
