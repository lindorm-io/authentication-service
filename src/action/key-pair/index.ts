export * from "./create-key-pair";
export * from "./expire-key-pair";
