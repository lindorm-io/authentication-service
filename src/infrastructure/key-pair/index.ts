export * from "./KeyPairCache";
export * from "./KeyPairRepository";
