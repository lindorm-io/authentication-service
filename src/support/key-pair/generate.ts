import {
  IGenerateECCKeysData,
  IGenerateRSAKeysData,
  KeyPair,
  KeyType,
  generateECCKeys,
  generateRSAKeys,
} from "@lindorm-io/key-pair";

export const generateKeys = async (type: KeyType): Promise<IGenerateECCKeysData | IGenerateRSAKeysData> => {
  switch (type) {
    case KeyType.EC:
      return generateECCKeys();

    case KeyType.RSA:
      return generateRSAKeys();

    default:
      throw new Error(`Wrong key pair type: ${type}`);
  }
};

export const generateKeyPair = async (type: KeyType): Promise<KeyPair> => {
  return new KeyPair(await generateKeys(type));
};
