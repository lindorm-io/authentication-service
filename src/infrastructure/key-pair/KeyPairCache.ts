import { schema } from "./schema";
import { CacheBase, ICache, ICacheOptions } from "@lindorm-io/redis";
import { IKeyPair, KeyPair } from "@lindorm-io/key-pair";

export interface IKeyPairCache extends ICache<KeyPair> {
  create(entity: KeyPair): Promise<KeyPair>;
  update(entity: KeyPair): Promise<KeyPair>;
  find(id: string): Promise<KeyPair>;
  findAll(): Promise<Array<KeyPair>>;
  remove(entity: KeyPair): Promise<void>;
}

export class KeyPairCache extends CacheBase<KeyPair> implements IKeyPairCache {
  constructor(options: ICacheOptions) {
    super({
      client: options.client,
      entityName: "KeyPair",
      expiresInSeconds: options.expiresInSeconds,
      logger: options.logger,
      schema,
    });
  }

  protected createEntity(data: IKeyPair): KeyPair {
    return new KeyPair(data);
  }

  protected getEntityJSON(entity: KeyPair): IKeyPair {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      algorithm: entity.algorithm,
      expires: entity.expires,
      passphrase: entity.passphrase,
      privateKey: entity.privateKey,
      publicKey: entity.publicKey,
      type: entity.type,
    };
  }

  protected getEntityKey(entity: KeyPair): string {
    return entity.id;
  }

  public async create(entity: KeyPair): Promise<KeyPair> {
    return super.create(entity);
  }

  public async update(entity: KeyPair): Promise<KeyPair> {
    return super.update(entity);
  }

  public async find(id: string): Promise<KeyPair> {
    return super.find(id);
  }

  public async findAll(): Promise<Array<KeyPair>> {
    return super.findAll();
  }

  public async remove(entity: KeyPair): Promise<void> {
    await super.remove(entity);
  }
}
