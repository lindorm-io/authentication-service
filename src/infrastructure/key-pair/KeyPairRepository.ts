import { IKeyPair, KeyPair } from "@lindorm-io/key-pair";
import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { MongoCollection } from "../../enum";
import { indices } from "./indices";
import { schema } from "./schema";

export interface IKeyPairFilter {
  id?: string;
  algorithm?: string;
  type?: string;
}

export interface IKeyPairRepository extends IRepository<KeyPair> {
  create(entity: KeyPair): Promise<KeyPair>;
  update(entity: KeyPair): Promise<KeyPair>;
  find(entity: IKeyPairFilter): Promise<KeyPair>;
  findMany(filter: IKeyPairFilter): Promise<Array<KeyPair>>;
  findOrCreate(entity: IKeyPairFilter): Promise<KeyPair>;
  remove(entity: KeyPair): Promise<void>;
}

export class KeyPairRepository extends RepositoryBase<KeyPair> implements IKeyPairRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.KEY_PAIR,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: IKeyPair): KeyPair {
    return new KeyPair(data);
  }

  protected getEntityJSON(entity: KeyPair): IKeyPair {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      allowed: entity.allowed,
      algorithm: entity.algorithm,
      expires: entity.expires,
      passphrase: entity.passphrase,
      privateKey: entity.privateKey,
      publicKey: entity.publicKey,
      type: entity.type,
    };
  }

  public async create(entity: KeyPair): Promise<KeyPair> {
    return super.create(entity);
  }

  public async update(entity: KeyPair): Promise<KeyPair> {
    return super.update(entity);
  }

  public async find(filter: IKeyPairFilter): Promise<KeyPair> {
    return super.find(filter);
  }

  public async findMany(filter: IKeyPairFilter): Promise<Array<KeyPair>> {
    return super.findMany(filter);
  }

  public async remove(entity: KeyPair): Promise<void> {
    await super.remove(entity);
  }
}
