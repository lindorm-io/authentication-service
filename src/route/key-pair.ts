import { BEARER_TOKEN_MW_OPTIONS } from "../config";
import { HttpStatus } from "@lindorm-io/core";
import { IKoaAuthContext } from "../typing";
import { Router } from "@lindorm-io/koa";
import { accountMiddleware } from "../middleware";
import { bearerTokenMiddleware } from "@lindorm-io/koa-jwt";
import { createKeyPair, expireKeyPair } from "../action";
import { clientMiddleware } from "@lindorm-io/koa-client";

export const router = new Router();

router.use(clientMiddleware());
router.use(bearerTokenMiddleware(BEARER_TOKEN_MW_OPTIONS));
router.use(accountMiddleware);

router.post(
  "/",
  async (ctx: IKoaAuthContext): Promise<void> => {
    const { type } = ctx.request.body;

    ctx.body = await createKeyPair(ctx)({ type });
    ctx.status = HttpStatus.Success.CREATED;
  },
);

router.patch(
  "/:id/expire",
  async (ctx: IKoaAuthContext): Promise<void> => {
    const { expires } = ctx.request.body;

    await expireKeyPair(ctx)({
      keyPairId: ctx.params.id,
      expires,
    });

    ctx.body = {};
    ctx.status = HttpStatus.Success.NO_CONTENT;
  },
);
