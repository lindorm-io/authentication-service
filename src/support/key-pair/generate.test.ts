import MockDate from "mockdate";
import { KeyType } from "@lindorm-io/key-pair";
import { generateKeyPair, generateKeys } from "./generate";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("@lindorm-io/key-pair", () => ({
  ...jest.requireActual("@lindorm-io/key-pair"),
  generateECCKeys: jest.fn(() => ({
    algorithm: "ec-algorithm",
    privateKey: "ec-privateKey",
    publicKey: "ec-publicKey",
    type: "ec-type",
  })),
  generateRSAKeys: jest.fn(() => ({
    algorithm: "rsa-algorithm",
    passphrase: "rsa-passphrase",
    privateKey: "rsa-privateKey",
    publicKey: "rsa-publicKey",
    type: "rsa-type",
  })),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("generateKeys", () => {
  test("should generate EC keys", async () => {
    await expect(generateKeys(KeyType.EC)).resolves.toMatchSnapshot();
  });

  test("should generate RSA keys", async () => {
    await expect(generateKeys(KeyType.RSA)).resolves.toMatchSnapshot();
  });
});

describe("generateKeyPair", () => {
  test("should generate EC key pair", async () => {
    await expect(generateKeyPair(KeyType.EC)).resolves.toMatchSnapshot();
  });

  test("should generate RSA key pair", async () => {
    await expect(generateKeyPair(KeyType.RSA)).resolves.toMatchSnapshot();
  });
});
