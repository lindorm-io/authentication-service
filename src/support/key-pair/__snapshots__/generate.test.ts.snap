// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateKeyPair should generate EC key pair 1`] = `
KeyPair {
  "_algorithm": "ec-algorithm",
  "_allowed": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_expires": null,
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_passphrase": null,
  "_privateKey": "ec-privateKey",
  "_publicKey": "ec-publicKey",
  "_type": "ec-type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`generateKeyPair should generate RSA key pair 1`] = `
KeyPair {
  "_algorithm": "rsa-algorithm",
  "_allowed": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_expires": null,
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_passphrase": "rsa-passphrase",
  "_privateKey": "rsa-privateKey",
  "_publicKey": "rsa-publicKey",
  "_type": "rsa-type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`generateKeys should generate EC keys 1`] = `
Object {
  "algorithm": "ec-algorithm",
  "privateKey": "ec-privateKey",
  "publicKey": "ec-publicKey",
  "type": "ec-type",
}
`;

exports[`generateKeys should generate RSA keys 1`] = `
Object {
  "algorithm": "rsa-algorithm",
  "passphrase": "rsa-passphrase",
  "privateKey": "rsa-privateKey",
  "publicKey": "rsa-publicKey",
  "type": "rsa-type",
}
`;
